### PR TITLE
created users_who_follow_vaults table

### PR DESCRIPTION
**Pull request by johnnieskywalker**
# [Store users who follow vaults in db](https://app.shortcut.com/oazo-apps/story/5897/store-followed-vaults-in-database-relation-1-address-to-many-cdpids)

  
## Changes 👷‍♀️
- added junction table which specifies which addresses follow which vaults of what type on given chainid
  
## How to test 🧪
- Run dev.sh
- see if migration works
- try to manually add in db records that define user following vault
- Check for edge cases like multiple users can follow the same vault of the same type
- One user can follow multiple vaults
- One user can follow vaults with the same ids on different chains

image

### CODE REVIEW

File: server/database/migrations/013-add-users-who-follow-vaults.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
+CREATE TABLE "users_who_follow_vaults" (
+    "user_address" TEXT NOT NULL,
+    "vault_id" INTEGER NOT NULL,
+    "vault_type" "vault_type" NOT NULL,
```
Comment (piotrkonowrocki) [inline server/database/migrations/013-add-users-who-follow-vaults.sql:4]: Why do we need `vault_type`, is it anyhow connected to the feature?

File: server/database/schema.prisma [hunk 2]
```diff
@@ -28,12 +28,14 @@ model TosApproval {
 }
 
 model Vault {
-    vault_id      Int       @unique
+    vault_id      Int                    @unique
     type          VaultType
     owner_address String
     chain_id      Int?
+    users         users_who_follow_vaults[]
```
Comment (piotrkonowrocki) [inline server/database/schema.prisma:35]: Is there any place where we need to display such information? I thought it's going to be based on simple queries from one table.

To see if user follows specific vault something like `SELECT * FROM users_who_follow_vaults WHERE vault_id == vaultId` AND user_address == userAddress`. Or instead of `SELECT` something like `COUNT`, because we only need confirmation if that record exist.

And to get list of all vaults followed by user for overview page `SELECT * FROM users_who_follow_vaults WHERE  user_address == userAddress`.

What am I missing?

File: server/database/schema.prisma [hunk 3]
```diff
@@ -45,13 +47,14 @@ enum VaultType {
...
+    address                    String                 @unique
+    timestamp                  DateTime               @default(now())
     user_that_referred_address String?
-    user_that_referred         User?         @relation("UserToUser", fields: [user_that_referred_address], references: [address])
-    referred_users             User[]        @relation("UserToUser")
+    user_that_referred         User?                  @relation("UserToUser", fields: [user_that_referred_address], references: [address])
+    referred_users             User[]                 @relation("UserToUser")
     weekly_claims              WeeklyClaim[]
     accepted                   Boolean
+    vaults                     users_who_follow_vaults[]
```
Comment (piotrkonowrocki) [inline server/database/schema.prisma:57]: Same as above - does this table really need to aggregate all followed vaults?

File: server/database/migrations/013-add-users-who-follow-vaults.sql (hunk 1 above)
Comment (johnnieskywalker) [reply server/database/migrations/013-add-users-who-follow-vaults.sql:4]: We'll need limit of 10 followed vaults per type as defined in here https://app.shortcut.com/oazo-apps/story/5893/validation-limit-number-of-followed-vaults

File: server/database/schema.prisma (hunk 3 above)
Comment (johnnieskywalker) [reply server/database/schema.prisma:57]: It doesn't it's just definitrion of relation so prisma would know which table is referred by other table, this syntax seem confusing I agree :shrug: 

https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#explicit-many-to-many-relations

File: server/database/migrations/013-add-users-who-follow-vaults.sql (hunk 1 above)
Comment (piotrkonowrocki) [reply server/database/migrations/013-add-users-who-follow-vaults.sql]: I spoke with Sam about it, he agreed that we shouldn't care about vault type and users can just hold up to 30 follows.

**Review by piotrkonowrocki — APPROVED** [review]
I reviewed and tested, I left one round of comments and concerns and now I think all works well, but I don't feel like database expert so I strongly suggest that it would be good idea if, for example @halaprix could review it as well.

**Review by halaprix — APPROVED** [review]
works as intended
image
